### ISSUE TITLE
custom dimensions and measures, table calculations ala Looker

### ISSUE THREAD

**Issue by djayatillake** [lightdash/lightdash#266]
https://docs.looker.com/exploring-data/using-table-calculations
https://docs.looker.com/exploring-data/adding-fields/custom-measure

I think a way this could be better done is for them to be defined in sql or some kind of natural language/GUI... Looker has opted for some horrible medium which is something like an excel function

**Comment by owlas** [comment]
Hey @djayatillake, glad you opened this one! We've been experimenting a little behind the scenes.

I think you're right that there's two workflows here: 

1. (**supported**) Ability to define custom metrics with raw sql (ultimate flexibility) inside dbt models
2. (**not supported**) Ability to perform custom table calculations in the UI


## Custom metrics with raw SQL (supported)

This is actually possible today and quite similar to looker: 

* **Aggregate metrics** accept a raw sql string and lightdash will then aggregate the result. e.g. `rawSql: col1 + col2` in a `sum` metric gives: `SUM(col1 + col2)`
* **Non-aggregate metrics** accept a raw sql string that lightdash expects to already be aggregated e.g. `rawSql: SUM(col1 + col2)` for a `number` metric.

There are even some calculations that typically get pushed to the UI (Looker would call them post-sql measures) but could actually done using analytic functions in sql.

A personal favourite is a lightdash metric for a cumulative sum:

```
columns:
  - name: sale_atm
    description: Total amount spent at checkout
    meta:
      metrics:

        # this metric will just be the sum of revenue
        total_revenue:
          type: sum

        # this metric will be a rolling sum of revenue
        cumulative_revenue:
          type: number
          rawSql: SUM(${total_revenue}) OVER(ROWS UNBOUNDED PRECEDING)
```

Pros: written as code, version controlled, etc.
Cons: hard to understand, still not accessible to non-technical users 


## Table calculations defined through the UI (not-supported)

So we still have the dilemma of how people that don't use SQL do calculations in the UI. I'm quite skeptical of many low-code sql builders. Looker's excel-style formulas is pretty common across tools.

We're still deciding what the right level of config should be in the UI. Today you can only create calculations as code in your dbt models. That may change in future

What are you thoughts? Should analysts be able to run calculations in the UI? Are there better alternatives to exposing excel-like or pseudo-sql calculations to non-technical users in the UI?

**Comment by djayatillake** [comment]
I think analysts do need to be able to do calculations in the UI, otherwise you end up having to build every variant they can think of and it stops them being able to self serve. Them being able to define a custom dimension (eg case when web_cohort < app_cohort then 'web first' else 'app first' end as journey), is really helpful to reduce what AE has to build in dbt to be minimum needed; in the example in brackets we would then only need to build the cohort fields.

Same for custom measures, there are one-time things that analysts or users sometimes want to count or sum that you don't always want to have to build in dbt each time.

Table calculations which are just one column + another one or one column / by another one are probably needed to be able to be defined in the UI, not necessarily calculated there. Certainly, as a starting point to simplify your UI you may want to push any such table calcs down to sql.

**Comment by owlas** [comment]
Some more thoughts about this:

* For now, I don't want to introduce any sort of DSL (like excel-esque) language for doing table calculations. We should just use SQL. If we need a UI to make it a bit simpler to build SQL (much like a lot of BI tools have done: chartio, preset, metabase) then we can add that. Just like you said @djayatillake - let's stick with SQL.
* adding custom metrics on the frontend should be relatively easy, we just need to include it when we run a query (i.e. run these metrics from the dbt schema.yml files and also this extra one I just wrote in the UI)
* saving custom metrics is trickier, and I think we should follow the same approach we discussed for the saved charts (@ZeRego):
  1. They **can** be written as code in you schema.yml file (this is already possible with non-aggregate metrics)
  2. They can **also** be written to the database (for non-technical users) where they will just get merged with existing metrics
  3. When you create a metric in the UI, you have the option to export it as `yml` and then you can include it in your dbt project manually. This saves the complexity of us actually writing to the dbt project ourselves.
* For "post-sql" metrics (like running totals) - I think this is something we should again solve at the sql level. My initial thought is every metric query runs inside a CTE and then you get the option to apply some post-sql metrics on top.

So tickets for this would be (in approximate order of first to implement to last to implement):
 - [ ] Allow users to define additional metrics in the UI. So you can test new metrics in the UI but they aren't persisted to your `schema.yml` or a database. They'll disappear when you close the app.
 - [ ] Allow users to copy the `.yml` for their custom metric to clipboard, then it can be added to their dbt project manually
 - [ ] Allow users to save custom metrics to the database
 - [ ] When loading an explore, load all metrics from the `schema.yml` file and also from the user's custom metric collection.

At this point we have everything for custom aggregate/non-aggregate metrics in the UI. But we still don't support "post-sql" metrics.

For "post-sql" metrics:

- [ ] Define some default post-sql metrics (like running total, etc.)
- [ ] Update SQL query builder to accept post-sql metrics
- [ ] Show post-sql metrics in the UI
- [ ] Allow users to provide post-sql metrics in their `schema.yml` ? not sure yet how this will work.

**Comment by djayatillake** [comment]
Replied in line:

> Some more thoughts about this:
> 
> * For now, I don't want to introduce any sort of DSL (like excel-esque) language for doing table calculations. We should just use SQL. If we need a UI to make it a bit simpler to build SQL (much like a lot of BI tools have done: chartio, preset, metabase) then we can add that. Just like you said @djayatillake - let's stick with SQL.

--yeah definitely stick with sql, most dbs out there support reference of a field defined in sql in the same select statement
--can see users needing help where it's a bit complicated like running totals, this can be achieved with window functions but can not straightforward for someone to write

> * adding custom metrics on the frontend should be relatively easy, we just need to include it when we run a query (i.e. run these metrics from the dbt schema.yml files and also this extra one I just wrote in the UI)

--exactly

> * saving custom metrics is trickier, and I think we should follow the same approach we discussed for the saved charts (@ZeRego):
>   

--I don't think you need to save custom metrics... otherwise you'll end up with metric spaghetti really fast. Looker doesn't save these... you can share a URL where someone has made it and then you can see it. I don't think any business user should be allowed to create a custom metric and save/share, they should be going to Analytics Engineers and BI and saying I've made this custom metric I want to share it or for it always to be there... then you can have the conversation to see if it's genuinely something new or something you want to adopt and maintain.

>   1. They **can** be written as code in you schema.yml file (this is already possible with non-aggregate metrics)
>   2. They can **also** be written to the database (for non-technical users) where they will just get merged with existing metrics
>   3. When you create a metric in the UI, you have the option to export it as `yml` and then you can include it in your dbt project manually. This saves the complexity of us actually writing to the dbt project ourselves.

-- this third option is interesting as it allows for that dialogue for a PR to happen where a business user wants a custom metric to persist but the AEs can say if it already exists or they don't want it to persist for some reason

> * For "post-sql" metrics (like running totals) - I think this is something we should again solve at the sql level. My initial thought is every metric query runs inside a CTE and then you get the option to apply some post-sql metrics on top.
> 
> So tickets for this would be (in approximate order of first to implement to last to implement):
> 
> * [ ]  Allow users to define additional metrics in the UI. So you can test new metrics in the UI but they aren't persisted to your `schema.yml` or a database. They'll disappear when you close the app.
> * [ ]  Allow users to copy the `.yml` for their custom metric to clipboard, then it can be added to their dbt project manually

--upgrades user story for this would be that it opens a PR with the relevant merge to .yml

> * [ ]  Allow users to save custom metrics to the database

--I don't think this is necessarily a good idea, ideally you want all metrics defined in one place... this is the joy of Lightdash that you are moving stuff from being defined in dbt and lookml/something else to just dbt

> * [ ]  When loading an explore, load all metrics from the `schema.yml` file and also from the user's custom metric collection.

--don't think this is a good idea for the reasons mentioned above

> 
> At this point we have everything for custom aggregate/non-aggregate metrics in the UI. But we still don't support "post-sql" metrics.
> 
> For "post-sql" metrics (less sure on this):
> 
> * [ ]  Define some default post-sql metrics (like running total, etc.)
> * [ ]  Update SQL query builder to accept post-sql metrics
> * [ ]  Show post-sql metrics in the UI
> * [ ]  Allow users to provide post-sql metrics in their `schema.yml` ? not sure yet how this will work.

I agree with this all being in sql even if you make a nice interface for non-tech users to abstract complex to build window functions by asking for running totals etc. The only thing I'm thinking is about runtime... adding a window function onto a query that scans a few hundred million rows may massively slow the query down whereas when it runs on the array of data returned in the UI (often just thousands of lines) it's very fast... can be microseconds. I understand this means storing the query results in something to then query but it's a trade-off... maybe one that you can make down the road but it would be interesting to see an example of extra cost/time from this trade-off. How crazy would it be for the instance that someone's Lightdash query is on to have it's own sqlite or some kind of mini-db to run those queries on the query results returned? Have you seen duckdb?

**Comment by owlas** [comment]
> I don't think you need to save custom metrics

I definitely see potential pitfalls with allowing users to save custom metrics. That said, a saved metric might not necessarily be available to the whole org. Just a specific user. It may be that a custom metric persists on a single explore but can't be "reused" across many easily (making not so delicious spaghetti).

> third option is interesting as it allows for that dialogue for a PR to happen

> opens a PR with the relevant merge to .yml

I think effort would be better spent on this workflow (above), where we make custom metrics easy to integrate into the dbt project. For two reasons:
1. Creates a great developer experience: metrics can be developed and tested in the Lightdash UI and then copied into the codebase. Rather than writing the yaml directly.
2. Allows non-technical users to contribute their experience but still have it captured in the core code base.

It would be great to have GitHub more tightly integrated to allow the user to manage their repo too.

> The only thing I'm thinking is about runtime... 

Yeah we have a bunch of other performance improvements to make on the query side (same for sorting and filtering, which can often be done in memory). 

I haven't seen duckdb! That's insanely cool. I'm going to do some digging. One potential problem I see is that the SQL may vary slightly depending on the target warehouse vs. duckdb.

Right now everything is sent to the warehouse for simplicity, consistency, and for having the most up to date data (so no caching worries). But the major downside is performance. And that does kinda suck. Because Lightdash is snappy except the queries make it feel slower.

**Comment by owlas** [comment]
@all-contributors add @djayatillake for design

**Comment by ZeRego** [comment]
+1 to not save these metrics in the DB right now. We can wait for community feedback and see if is really necessary.

**Comment by TuringLovesDeathMetal** [comment]
+1 as well. I don't think you should be allowed to save these metrics directly. _**love**_ the idea of being able to copy the SQL that generates the metrics though so you can add it to your .yml files if you think it's something worth persisting.

The post-sql stuff sounds really great too! I have some ideas for the first few default post-sql metrics I'd love to see 😏 

>The only thing I'm thinking is about runtime... adding a window function onto a query that scans a few hundred million rows may massively slow the query down whereas when it runs on the array of data returned in the UI (often just thousands of lines) it's very fast... can be microseconds.

^This is really interesting! I might be totally missing something here, but could we not just apply these post-SQL transformations to a CTE that's only a tiny subset of the data? 

We add a `limit` clause at the end that reflects the number of rows you've selected in the UI (default `LIMIT 500`), so could we make this "first chunk of SQL" into a CTE, then any post-SQL transformations are applied to this subset of data?

E.g. 
```
WITH stg1 AS (
SELECT
  payments.payment_id AS "payments_payment_id",
  SUM(payments.amount) AS "payments_total_revenue"
FROM "demo_db_57w5"."jaffle"."payments" AS payments
LEFT JOIN "demo_db_57w5"."jaffle"."orders" AS orders
  ON (orders.order_id) = (payments.order_id)
LEFT JOIN "demo_db_57w5"."jaffle"."customers" AS customers
  ON (customers.customer_id) = (orders.customer_id)

GROUP BY 1
ORDER BY 1

LIMIT 500
)

SELECT
*,
SUM(payments_total_revenue) OVER(PARTITION BY NULL ORDER BY payment_id) AS payments_total_revenue_running_total

FROM stg1
```

**Comment by owlas** [comment]
I guess the problem is that it will still rerun the CTE (which might contain a huge scan of data, think bigquery costs) everytime you change or add a post sql measure. 

Ideally (for cost and speed) the CTE result would be cached.

**Comment by djayatillake** [comment]
If you hadn't already run the query before wanting to do the table calc then the CTE way would work, but from experience this almost never happens; users prefer to see the data first before making table calcs.

Are you sure you would want to cache the query result in your infra instead of storing it as a temp table in the users DWH that you can then query as much as you like in that session?

**Comment by owlas** [comment]
> users prefer to see the data first before making table calcs.

Definitely!

> sure you would want to cache the query result in your infra

Not sure. Let's move this to another issue

**Comment by TuringLovesDeathMetal** [comment]
ah, that makes a lot of sense - yep, totally see the problem now. Thanks for clarifying ❤️

**Comment by owlas** [comment]
Now to the nitty gritty.

## Requirements

I think we're all agreed:

* Ability to add a new column in the UI that is derived from existing data in the table
* e.g. I can add two columns together to create a third column
* e.g. I can compute a running total of another column
* They shouldn't be persisted as a "new metric" available for other users
* They should be persisted if I save the query or chart containing this custom metric (i.e. if I share it with you, you'll see my custom column)

## Implementation

There are two quite different implementations to achieve this:

### Background

To understand why these are different, here's a reminder of how queries run on Lightdash:

1. You write some SQL templates in dbt `meta` to define metrics and dimensions
2. Lightdash runs a compile step. This translates the SQL templates into raw sql, which will be used later for queries
3. User selects metrics, dimensions, filters, sorts, limits. Remember the key thing here is that these are "composable". So you can select any combo of metrics and dimensions. That's why things like running totals don't work, because they rely on the data being a certain shape (i.e. you have to have the table sorted properly for the metric to be right, bad)
4. Lightdash takes the compiled explore (from 2) and the selected metrics/dimensions/sorts (from 3) and creates a SQL string (we call this a metric query)
5. The SQL string is sent to the data warehouse to get results

### The two options

1. Metrics created in the UI are no different to existing metrics. They're added to the explore (on the fly), the explore is recompiled. And then the query runs with the custom metric.
2. Metrics created in the UI are different from existing metrics and are run only after the main query has run (in reality we run the main query in a CTE then the custom metrics are added).


### Option 1. Table calculations are the same as existing metrics

In the UI you would add a new column and write a small sql string to specify how the new column is calculated from existing metrics. This is identical to a non-aggregate metric. What we'd do here is take the metric definition from the UI, add it to the explore and recompile it (step 2) above, the rest of the query would just follow as normal.

Pros:
* we don't introduce any new entity to Lightdash
* we can have a feature where you copy the `.yml` for the metric, and later add it to your project

Cons:
* Post-sql metrics (like running totals) are kind of a hack here, because they aren't composable like the rest of the metrics, they only work for a specific results set

### Option 2. Table calculations are a different type of metric

In the UI you would add a new column and write a small sql string just like before. But instead of creating a new metric, we would change the MetricQuery (step 4 above) to accept a list of these table calculations. These would get compiled when you run the query. This feels more natural to me.

Pros:
* table calculations have their own entity - which makes them distinct
* they will handle things like running totals well (we could even have checks that certain base metrics + dimensions exist in the results table in order to compute the table calculations)

Cons:
* The copy to `.yml` feature makes less sense, as these aren't metrics

**Comment by owlas** [comment]
I prefer option 2, which I think is more true to what the user expects and also how the UI will behave. It would also formalise a way for the API to expose these custom metrics too (might be useful in future)

**Comment by owlas** [comment]
@ZeRego made the point that these aren't mutually exclusive - that's true!

**Comment by ZeRego** [comment]
I see each option as a separate feature. 
I can also see us supporting saved charts in yml, which would include these table calculation.

**Comment by mike-weinberg** [comment]
I find myself agreeing with @TuringLovesDeathMetal and @owlas - I imagine for this a ux in which creating a custom metric automatically creates a branch (which is automatically maintained up to date with master) and a pull request which adds the new metric to the yml for the explore. From the user's perspective, they get to play with the new metric. This would close the loop between the dashboard user/developer and the analytics engineer(s) by replacing what might previously have been a jira ticket with what is now an honest to goodness PR.

**Comment by djayatillake** [comment]
Would you want every custom metric created to branch/pr? Often custom metrics and dimensions are throwaway, someone makes one graph with them, maybe shares the graph with others and then those metrics and dimensions are not used again; feels like it should be if a user wants a custom metric to persist that you go down the branch/pr route.

**Comment by TuringLovesDeathMetal** [comment]
I might have added some confusion in here! 

I _**don't**_ think that we should create branches/PRs for each time we add a custom metric to a visualization. 

I _**do**_ think that there should be an easy way to add a custom metric you've made into your dbt project.

I imagine a world where I've created some super awesome custom metric in the Lightdsh UI, and I want to add it to my dbt project so that it's more permanent/everyone else has access to it when they're doing analyses. 

So, with my custom metric, I can click on some button that says `show me the YAML` (maybe a voice over of [show me the money](https://www.youtube.com/watch?v=mBS0OWGUidc), but not a V1 requirement...). Then, I can easily add this to dbt schema file, and create a PR to merge this.
